### ISSUE TITLE
Pixel format API design

### ISSUE THREAD

**Issue by ids1024** [rust-windowing/softbuffer#98]
We need an API to list available pixel formats, and select one. Some considerations:
* We want a way to support transparency: https://github.com/rust-windowing/softbuffer/issues/17
* The web backend current converts BGRX to RGBA. It would be better if it could use RGBA directly, if an application can render that way.
* It looks like for https://github.com/rust-windowing/softbuffer/issues/44 we'd have to use RGBX, rather than BGRX.
* For https://github.com/rust-windowing/softbuffer/pull/95, it seems BGRA is supported, but not BGRX. So we need to make sure to set alpha to 255.
* Wayland's `wl_shm` can support many formats, but only two are required to be supported by compositors.

So some questions here:
* What formats should be supported?
   - We want to offer a format for each platform that can be used efficiently without conversion, at least, and there doesn't seem to be one that can be assumed everywhere, unfortunately.
   - There are *many* formats that could be supported on at least some platforms.
* Do we want to support formats that are not 4 byte per pixel? This will impact the API, which currently uses u32 values for pixels.
* How to we communicate what formats are supported, and what formats are optimal?
* We still want at least one format to work on all platforms, even if it requires conversion on some, right? What formats should be universally supported, then?

**Comment by notgull** [comment]
[Here](https://github.com/bread-graphics/genimage/blob/master/src/format.rs) is an API I wrote a while back for an image crate I was writing (abandoned as of now). The idea is that is specifies RGBA channels, their sizes, their order and whether or not they are present. It's based on the setup that `pixman` uses. I think a similar type would be good for `softbuffer`.

What if we made it so, no matter what format the buffer is, it's transcoded to the proper format? However, the API would expose a "transcode-free formats" function (maybe an `Iterator`?) that indicates which formats aren't transcoded at all?

**Comment by ids1024** [comment]
Technically that doesn't cover every possible format. Wayland for instance [supports](https://wayland.app/protocols/wayland#wl_shm:enum:format) things like YUV, indexed color, etc. But I doubt YUV is useful for softbuffer consumers, and indexed color (hopefully) isn't relevant these days, nor is it useful without knowing the palette. :laughing:

I guess that iterator would be returned by a method of `Surface`? And there would be a `set_format` method to set the format for the next `buffer_mut`? Unlike `resize` this call would be optional, and default to the current default.

But still, we have the fact `Buffer` derefs to `&[u32]`, which is a problem if we want to support formats with more or less than 32 bits per pixel.

**Comment by ids1024** [comment]
The most obvious way to specify formats would be to provide an `enum` with every format supported by at least one backend. Regardless, backends likely use enums, and need to be mapped that way.

Of course we also need to handle conversion some way. Which is simple for variations on RGBA in different orders.

Transparency support is somewhat special. We can't really convert RGBA to RGBX and expect that to work as expected. The application needs to know if transparency is supported. Also a particular backend might only support RGBA here, but have a hint to indicate that the surface is opaque. This seems to be the case with `IOSurface` an `CALayer`'s `isOpaque` property (no idea if that will have any impact when using an `IOSurface`).

**Comment by daxpedda** [comment]
> Transparency support is somewhat special. We can't really convert RGBA to RGBX and have that to work as expected (but can do the reverse). The application needs to know if transparency is supported. Also a particular backend might only support RGBA here, but have a hint to indicate that the surface is opaque. This seems to be the case with `IOSurface` an `CALayer`'s `isOpaque` property (no idea if that will have any impact when using an `IOSurface`).

I agree with the `enum` idea. Couldn't we encode alpha support into the texture format `enum`? So `BGRX` and `BGRA` could just be two different formats.

**Comment by daxpedda** [comment]
> The complication there is format conversion.

I'm in favor of doing no conversion at all. We take what the user supplies and that's it. The user has to supply it correctly.

> So do we not allow that conversion, and require an application to check for a format with alpha before trying to use transparency?

One way I imagine we could "require to check the format" could be to make the format part of the buffer, basically:
```rust
enum Buffer<'a> {
	Bgrx(BgrxBuffer<'a>),
	Bgra(BgraBuffer<'a>),
	Rgba(RgbaBuffer<'a>),
}
```
I'm don't think I'm in favor of this API, because the API of each of these types would be exactly the same. Having a `format` function that returns the an `enum` with the format would make more sense to me.

In any case, I'm in favor of not doing any conversions at all.

**Comment by ids1024** [comment]
Not doing conversions makes things simpler, definitely. The trouble there is if there is no format that is supported on all backends. If we don't do conversion, does that mean the `web` backend would only advertise `RGBA`, and an application could only work on that backend if it can render to that (which may or may not be practical), or does the conversion itself?

So this compromises the goal of making `softbuffer` an easy way to put pixels on the screen across platforms. And we'd want to document the minimum set of formats a portable application needs to handle.

We could also provide helpers or a wrapper to do format conversion. At least we'd need to make the examples demonstrate what is needed for a portable application.

**Comment by daxpedda** [comment]
> If we don't do conversion, does that mean the `web` backend would only advertise `RGBA`, and an application [..] does the conversion itself?

Yes, that is what I would be in favor of.

> So this compromises the goal of making `softbuffer` an easy way to put pixels on the screen across platforms. And we'd want to document the minimum set of formats a portable application needs to handle.
> 
> We could also provide helpers or a wrapper to do format conversion. At least we'd need to make the examples demonstrate what is needed for a portable application.

Maybe I'm thinking about this wrong. The goal is to create a single portable API. But the reason we are here discussing different formats, is because not all backends work the same and there are differences that we can't bridge without introducing overhead. That is the reason why I want to have conversions not happen in `softbuffer`, but by the user, to enable a way to not pay the overhead in return for a non-portable API.

So my thinking now was instead of introducing helpers, maybe it would make more sense indeed to do the conversion in `softbuffer`, but introduce extension traits to circumvent that conversion by having to supply specific color formats.

Alternatively, we could also introduce `from_rgba()`, `from_bgrx()`, `from_bgra()` and have a function returning the format which does no overhead.

To summarize my braindump here:
- (non-portable) Return an `enum` `Buffer` type with a type for each format (https://github.com/rust-windowing/softbuffer/issues/98#issuecomment-1507456058).
- (non-portable) Provide a `format()` function telling the user which format needs to be supplied to the buffer, doing no conversions on the `softbuffer`-side. Optionally provide helper functions to do the conversion on the user side.
- (portable) Provide `from_rgba()`, `from_bgrx()`, `from_bgra()` methods, necessary conversions are done on the `softbuffer` side. If the user wants to circumvent the conversion overhead, a `format()` function can inform the user which format is native.

All APIs suggested here would be available on all platforms, non-portable just means the user has to handle multiple ways to draw the image to handle different formats for different platforms.

Currently I'm in favor of the last one.

**Comment by daxpedda** [comment]
https://github.com/rust-windowing/softbuffer/issues/98#issuecomment-1504259143
> What if we made it so, no matter what format the buffer is, it's transcoded to the proper format? However, the API would expose a "transcode-free formats" function (maybe an `Iterator`?) that indicates which formats aren't transcoded at all?

Basically is the conclusion I came to after a bunch of back and forth here (just an `enum` instead of an `Iterator`).

Which basically still leaves open the two problems posed by @ids1024.
- How to handle non `&[u32]` formats.
- How to handle transparency conversions.

I think the first problem can be addressed by providing conversion free methods. Alternatively the "multiple `Buffer` types" suggestion would address this to.

Transparency conversion is problematic, but we could apply the "pre-multiplied alpha" strategy, which is the best I can think of. Alternatively, again, could be addressed with the "multiple `Buffer` types" suggestion.

**Comment by ids1024** [comment]
Making `Buffer` an `enum` makes a certain amount of sense. But I think we want an API that lets the caller choose between multiple different formats out of those that are available on the platform, so it would just be a bit redundant. (If I'm setting the format, I'd match on the `Buffer` with a `unreachable!()` for types other than the one I selected.)

Having the core API always use a native format, and then providing some sort of helper for conversion, seems better now that I think of it. "Explicit is better than implicit", and done well that should still provide an easy to use API.

Though that still needs a good API. A `from_rgba()` method would probably still involve copying even if no conversion is needed, right? It would be good for it to still be no-copy where possible. I guess an API could wrap `Buffer`, also be derefable, and if conversion is needed, either allocate a `vec!` for the format converted from, or convert in place on present? (In-place conversion would clash with https://github.com/rust-windowing/softbuffer/issues/90).

**Comment by ids1024** [comment]
If we're unsure about non-u32 formats, we can leave those out for now, and update the ABI for that later (possibly in a breaking way).

I'd definitely say the priorities are:
* Being able to use a native format without conversion on all platforms, and without copying.
* Support for transparent windows where possible.

I'd like to fix those and get them in a tagged release. Those seem like the main limitations of softbuffer currently. https://github.com/rust-windowing/softbuffer/issues/39 and https://github.com/rust-windowing/softbuffer/issues/90 could be added with non-breaking changes, as an enhancement.

For most purposes on current platforms, you don't need pixel formats smaller than 32-bit. And higher bit depth (for things like HDR) is still niche and probably not something that you'd use a software renderer for.

**Comment by daxpedda** [comment]
> Though that still needs a good API. A `from_rgba()` method would probably still involve copying even if no conversion is needed, right?

True, I didn't think this through.

> I guess an API could wrap `Buffer`, also be derefable, and if conversion is needed, either allocate a `vec!` for the format converted from, or convert in place on present? (In-place conversion would clash with #90).

I'm not sure I properly understand how #90 plays a role here, could you elaborate? My first thought was to do in-place conversion on present.

I was also just thinking that reading pixels will be pretty weird as well, because it might not be the format users are using to write.

---

Maybe I was thinking about this wrongly, but it might be better to decide on the texture format on `Surface` creation, conversion would then happen in present if the format is not native and the buffer was changed. This would make reading and writing pixels use the same format as well. In addition, whatever API we will come up with to let the user use a desired format on the `Buffer`, it won't make a whole lot of sense, the most common use-case is to use a single texture format, native or not, it's unlikely anybody wants to switch formats between frames.

**Comment by ids1024** [comment]
> I'm not sure I properly understand how https://github.com/rust-windowing/softbuffer/issues/90 plays a role here, could you elaborate?

The goal of buffer age is to allow redrawing only part of the buffer, because the buffer matches what you presented `age` frames ago. If it were converted in place, the components would be mixed up, unless `buffer_mut` also converts the buffer back.

So I guess the format conversion wrapper could provide an `age` that passes through to the backend when conversion isn't needed, but always returns `None`/`0` for converted formats if conversion is done in place, or returns `1` if a separate vec were used.

> I was also just thinking that reading pixels will be pretty weird as well, because it might not be the format users are using to write.

Basically the same issue as buffer age. So if we want to support it, it would have to convert back. Which couldn't be done in-place for a front buffer that the display server is also accessing.

> Maybe I was thinking about this wrongly, but it might be better to decide on the texture format on Surface creation

Can we assume the formats supported by the display are the same for each surface, in all backends? Then `softbuffer::Context` could have a method to get a list of supported formats. Otherwise we need a `RawWindowHandle` or a `softbuffer::Surface` to get the supported formats, which is important for an application that could handle multiple formats depending on what the backend supports.

**Comment by daxpedda** [comment]
> Can we assume the formats supported by the display are the same for each surface, in all backends? Then `softbuffer::Context` could have a method to get a list of supported formats. Otherwise we need a `RawWindowHandle` or a `softbuffer::Surface` to get the supported formats, which is important for an application that could handle multiple formats depending on what the backend supports.

We have access to the `Surface` already when creating `Context`, right? So this shouldn't be a problem, but I don't know any of the details here. On Web there is only one supported format :laughing:.

**Comment by daxpedda** [comment]
> Creating a `Context` requires a `RawDisplayHandle`, while creating a `Surface` requires a `RawWindowHandle`.

Ah, I mixed it up, we have access to the `Context` when creating a `Surface`. So what you are saying is that on Wayland we need the `RawWindowHandle` to get the preferred format.

This is possible but awkward, to get the preferred format for a surface we need to have `RawWindowHandle`, which we then need to pass again when creating the `Surface` from the `Context`.

Maybe the solution here could be to create a `SurfaceBuilder` with `RawWindowHandle`, then decide on the format, which then creates the actual `Surface`.

> [..], and can change dynamically: https://wayland.app/protocols/linux-dmabuf-unstable-v1#zwp_linux_dmabuf_v1:request:get_surface_feedback

Uhm, that sounds very weird. Maybe we just want to document that and not handle it otherwise? Or is this a common scenario?

**Comment by ids1024** [comment]
We're not currently using that API for the Wayland backend, we're currently using `wl_shm` where there are no differences between surfaces. This would be if we wanted to use a `dmabuf` referencing an on-GPU buffer instead. Which might perform better at least under certain circumstances (presumably it's better on a unified memory architecture, at least, saving a copy?).

So I don't know that this is an issue with any current platform, but it's at least a potential concern, that could limit changes to backends or future backends.

Yeah, a `SurfaceBuilder` might be a good idea. I was thinking of having a default format, and a way to change it, but if the conversion is done by a separate helper that wouldn't really work so well since the default would be different on different platforms.

**Comment by madsmtm** [comment]
To move forwards on this a bit, I've tried to summarizes the natively supported formats on various platforms. Let me know if something is wrong, then I'll update it (or edit the comment yourself if you're a maintainer).

Formats are in `XYZW<bitsize>`, where `X` is byte 0, `Y` is byte 1 etc. in an `[u8]` buffer [(i.e. big-endian)](https://github.com/rust-windowing/softbuffer/issues/109). There's often support for a lot more than shown here, but I've extracted the interesting based on the ones the [`rgb v0.8.90-alpha.2`](https://docs.rs/rgb/0.8.90-alpha.2/rgb/index.html) crate.

Alpha means `0` is transparent, `255` is opaque.


## CoreGraphics

[Supported Pixel Formats](https://developer.apple.com/library/archive/documentation/GraphicsImaging/Conceptual/drawingwithquartz2d/dq_context/dq_context.html#//apple_ref/doc/uid/TP30001066-CH203-BCIBHHBB).

Supported on both macOS and iOS (might be more nowadays, the above link is dated):
- `Gray<u8>`
- `Rgb<u8>X`, `XRgb<u8>`, `Bgr<u8>X` and `XBgr<u8>` (24-bit / padded).
- `Rgba<u8>`, `Argb<u8>`, `Bgra<u8>` and `Abgr<u8>` (alpha channel might need to be pre-multiplied).


## Web

[`ImageData.data`](https://developer.mozilla.org/en-US/docs/Web/API/ImageData/data).

Supported:
- `Rgba<u8>`.

(Might be possible to hackily [convert the colors using filters](https://stackoverflow.com/a/43960991)).


## Android

[AHardwareBuffer_Format](https://developer.android.com/ndk/reference/group/a-hardware-buffer#ahardwarebuffer_format).

Supported:
- `Gray<u8>` (`AHARDWAREBUFFER_FORMAT_R8_UNORM`).
- `GrayA<u16>` (`AHARDWAREBUFFER_FORMAT_R16G16_UINT`).
- `GrayA<u8>` (`AHARDWAREBUFFER_FORMAT_R16_UINT`).
- `Rgba<f16>` (`AHARDWAREBUFFER_FORMAT_R16G16B16A16_FLOAT`).
- `Rgba<u8>` (`AHARDWAREBUFFER_FORMAT_R8G8B8A8_UNORM`).
- `Rgb<u8>X` (`AHARDWAREBUFFER_FORMAT_R8G8B8X8_UNORM`) (32-bit / padded).
- `Rgb<u8>` (`AHARDWAREBUFFER_FORMAT_R8G8B8_UNORM`) (24-bit / packed).


## Wayland

[`wl_shm::Format`](https://docs.rs/wayland-client/latest/wayland_client/protocol/wl_shm/enum.Format.html).

Supported by all renderers:
- `Bgra<u8>` (`argb8888` is little-endian)
- `Bgrx<u8>` (`xrgb8888` is little-endian)


## Orbital

[`Color::rgba`](https://docs.rs/orbclient/0.3.48/src/orbclient/color.rs.html#25-29).

Supported:
- `Bgra<u8>`.


## Windows

[`BITMAPINFOHEADER`](https://learn.microsoft.com/en-us/windows/win32/api/wingdi/ns-wingdi-bitmapinfoheader).
[`BITMAPV4HEADER`](https://learn.microsoft.com/en-us/windows/win32/api/wingdi/ns-wingdi-bitmapv4header) (if using alpha channel).

Supported:
- Any configuration of `A`, `R`, `G` and `B` (`Argb<u8>`, `Rgba<u8>`, `Abgr<u8>`, ...).
    - Though unsure if some of these will fail at runtime / be less performant?
- Probably also `Gray<u8>` and `GrayA<u8>`, though unsure here.


## X11

Unsure? Seems like it supports a lot, but are there any that you could consistently rely on always being available?


## DRM / KMS

[`DrmFourcc`](https://docs.rs/drm-fourcc/latest/drm_fourcc/enum.DrmFourcc.html).

A bunch are supported, probably all we'll ever need, though unsure if all hardware supports all of these.

**Comment by madsmtm** [comment]
Another option would be to do something like:
```rust
#[cfg(target_family = "wasm")]
type Pixel = rgb::Rgba<u8>;
#[cfg(not(target_family = "wasm"))]
type Pixel = rgb::Bgra<u8>;
```

The downside here is that it makes it a bit more difficult to write platform, but maybe that's a tradeoff that we're willing to accept? `Rgba<u8>` and `Bgra<u8>` do have really similar API surfaces.

Yet another option would be to define a pixel type ourselves:
```rust
#[repr(C, align(4))]
struct Pixel {
    #[cfg(target_family = "wasm")]
    red: u8,
    #[cfg(target_family = "wasm")]
    green: u8,
    #[cfg(target_family = "wasm")]
    blue: u8,

    #[cfg(not(target_family = "wasm"))]
    blue: u8,
    #[cfg(not(target_family = "wasm"))]
    green: u8,
    #[cfg(not(target_family = "wasm"))]
    red: u8,

    alpha: u8,
}
```

That should allow us to make it clear which conversions are valid cross-platform, though I think there'd be value in at least providing conversion methods to/from the types in the `rgb` crate.

(I should say: I believe it's much more valuable to support a single pixel format well, than support a bunch of different formats).

**Comment by ids1024** [comment]
> would it be possible to query the available formats, and then falls back to copying when not available? Or is that not an option without interfacing with the event loop?

Supported shared-memory formats on Wayland are global (not tied to a specific window) so we could block on a server round-trip in `Context::new()` to get the list of supported formats. Then either `Context` or `Surface` can have a method to query supported formats without blocking.

**Comment by ids1024** [comment]
I guess I'm misunderstanding what `XShmPixmapFormat` means; that seems to be just either `XYPixmap` or `ZPixmap`. And should always be `ZPixmap` on modern systems that don't use seperate bitplanes for different colors?

So I guess the format we can use on X11 is defined by visuals instead. But when I run `xdpyinfo`, I seem to just see visuals with `red, green, blue masks:    0xff0000, 0xff00, 0xff`. So I think that means we only have one option for byte order of the pixels? (Though the protocol specification could allow for multiple; the implementation doesn't seem to.)

You can see this in softbuffer by swapping `visual.red_mask` and `visual.blue_mask` when it's matching visuals.

So if I understand that correctly, trying to require RGBX would result in always falling back to the slow copy path on X11?

On Wayland, we can probably expect the RGBX ordering like this to be supported, in most compositors. (Although BGRX is the only format guaranteed by the protocol.)

**Comment by ids1024** [comment]
Oh, that might be. I'd also need to test it to see if that could work on Windows. The win32 API documentation doesn't seem very clear, in general.
